### ISSUE TITLE
Document Provider Implementation

### ISSUE THREAD

**Issue by Nutomic** [syncthing/syncthing-android#205]
https://developer.android.com/guide/topics/providers/document-provider.html

Requires syncthing/syncthing#723

**Comment by calmh** [comment]
Does it though? Couldn't you implement a document provider that provides access to the files locally - you know where they are, because you told syncthing-core about them? (If we've been over this already, sorry, I may have forgotten or missed the point... :/ )

**Comment by Nutomic** [comment]
Ah yeah that's right, didn't think if it right now. 

Now I just need to find the time for this D:

**Comment by Nutomic** [comment]
Ah @calmh, the actual point of this is that you can access files remotely. So you don't have to sync your 10 GB of music to your phone just to listen to a single song.

It looks like `/rest/tree` is already a first step to this, what's missing (for inital read-only support) is file download from the global model.

**Comment by AudriusButkevicius** [comment]
Well, this should be handled with selective sync or ignores surely?

**Comment by Nutomic** [comment]
It's a bit different than that, because you can access *any* file in the share without selecting subfolders beforehand. This is implemented by Google Drive as a "server". For "clients", I only know DocumentsProvider in OmniROM, which is a file manager that works on local and remote files seamlessly.

I agree that selective sync would also be useful (but I understand you want to rewrite ignores soon).

**Comment by Nutomic** [comment]
Here's how this looks:
[Select which files to show](http://i.imgur.com/4eQZOSv.png)
[Display folders](http://i.imgur.com/Hq3BPAa.png) This would list syncthing shares first, and then let you browse/view/edit syncthing files.

**Comment by rajsinghUSA** [comment]
Is this feature pretty dead in the water right now? It's the only thing keeping me from shifting over to Syncthing.

**Comment by Nutomic** [comment]
No one is working on it right now. But contributions are always welcome :)

**Comment by sqrt-1764** [comment]
What are the exact features the document provider should publish?
If it is to publish the content of the shared folders or receive files for those folders, I think I could handle that and might be interested in implementing this.

Regarding selective/ondemand download: Someone would have to point me to the hook into Syncthing where a document provider could hook into. Currently I haven't taken such a close look at the code ...

**Comment by capi** [comment]
@sqrt-1764 At the moment the main problem will be that the native Syncthing binary offers no way of a "virtual filesystem" where we could hook the document provider into.

**Comment by Nutomic** [comment]
Exactly, but this might be done soon when SyncthingFUSE is merged into Syncthing in some way.

https://forum.syncthing.net/t/syncthingfuse-first-release-candidate/6402/2
https://forum.syncthing.net/t/dropbox-infinite/7237

**Comment by sqrt-1764** [comment]
Hmm, I just took a quick glance in the REST-API. From there one should get the info which files reside in which directories - right?
So this should be all the info a document-provider service needs to publish the files to the Android-System. Sounds doable ...

**Comment by Nutomic** [comment]
The problem is we can't get the file contents yet, or change the files. That still needs support in Syncthing.

**Comment by AudriusButkevicius** [comment]
Unless like the first comment already explains, you already have the files on disk?
This is not about fetching files on the fly, this is about implementing the provider.

**Comment by Nutomic** [comment]
That's right. But as I understand it, document providers are meant to access remote files. Local files can already be accessed, so there's no point in doing that again.

**Comment by sqrt-1764** [comment]
In the case of local files Document-Providers enable the apps to those files without the need to use the permission to read and or write to the external storage. As the user selects the files / folders manually Android grants the right to acess the file to the calling app.
So you don't need global read or write access to the local files in order to access the contents of the document-provider.

**Comment by Nutomic** [comment]
But the documents provider already gives access to all local files by default (you might have to toggle a checkbox to show the SD card, but that's it.

**Comment by sqrt-1764** [comment]
That's correct.
A document-provider from Syncthing at the moment would only present the files in a different way (ordered by the shares in Syncthing regardless where on the device those shares reside).

**Comment by Nutomic** [comment]
Yeah, that seems quite pointless seeing as we'll have to implement it again anyway.

**Comment by rajsinghUSA** [comment]
Any updates on on-demand file access for android?

**Comment by Nutomic** [comment]
There's a new app with that functionality, check the forum.

**Comment by rajsinghUSA** [comment]
@Nutomic any timeline to implement this functionality in the official app? It's really an indispensable feature for a phone.

**Comment by Nutomic** [comment]
This is not supported by Syncthing, so it cant be implemented (and there are no active plans to implement it in Syncthing right now). You can use [a-sync-browser](https://play.google.com/store/apps/details?id=it.anyplace.syncbrowser) instead.
